#### ✅ 1. What is an API?
API = Application Programming Interface

An API allows two software applications to communicate with each other.

You use APIs every day:

Example	Behind the Scenes
Google Maps in a mobile app	App sends API request → Google sends map data

Weather app	Your phone uses weather API to get temperature

Online shopping	Fetching product list from Amazon API

Simple Analogy


### Think of a waiter in a restaurant:

You (client) → send request

Waiter (API) → carries request to kitchen

Kitchen (server) → processes request

Waiter returns food (response)

#### 👉 API = messenger

🟦 2. HTTP Methods (GET, POST)

APIs use HTTP methods to perform actions.

####🔹 GET Request

Used to fetch data




#### Example:

Get weather

Get user information

GET https://api.example.com/users

#### 🔹 POST Request

Used to send or create data

Example:

Register a user

Submit a form

POST https://api.example.com/login

##### 🟦 3. URLs and Endpoints

An API endpoint is a specific URL that gives access to specific data.

Structure:
Base URL:      https://api.example.com
Endpoint:      /users
Full URL:      https://api.example.com/users

Example:
https://jsonplaceholder.typicode.com/posts


This endpoint returns a list of posts.

🟩 PART 2 — Python Requests Module (40 minutes)
🟦 4. Installing requests

requests is a popular Python

| Method | Typical use     | Example (conceptual)                                   |
| ------ | --------------- | ------------------------------------------------------ |
| GET    | Read data       | Get list of posts: GET /posts.freecodecamp​            |
| POST   | Create new data | Add a post: POST /posts with JSON body.freecodecamp+1​ |

#### Part 4: Installing requests in Python
This is the bridge from theory to code.

requests is a popular third‑party Python library to make HTTP requests easily.​

It is not part of the Python standard library; you must install it with pip.​

Command (tell them to run in terminal or VS Code integrated terminal):

In [ ]:
!python -m pip install requests


In [ ]:
import requests


### First GET request example
Use a real, safe, public API such as JSONPlaceholder.

##### Example 1: Basic GET, status code
Explain:

requests.get(url) sends a GET request.​

The result is a Response object, with status_code, text, .json(), etc

In [1]:
import requests

url = "https://jsonplaceholder.typicode.com/posts"  # public fake API
response = requests.get(url)

print("Status code:", response.status_code)
print("Type of response object:", type(response))


Status code: 200
Type of response object: <class 'requests.models.Response'>


### Discussion:

#### 200 status means “OK” (success).​

If status is 404 → resource not found; 500 → server error.​

Ask: “What status code did you get?” and “What might it mean if we get 404?”

In [2]:
import requests

url = "https://jsonplaceholder.typicode.com/posts/1"
response = requests.get(url)

# Convert JSON response to Python data (dict)
data = response.json()

print("Type of data:", type(data))
print("Data:", data)
print("Title field:", data["title"])


Type of data: <class 'dict'>
Data: {'userId': 1, 'id': 1, 'title': 'sunt aut facere repellat provident occaecati excepturi optio reprehenderit', 'body': 'quia et suscipit\nsuscipit recusandae consequuntur expedita et cum\nreprehenderit molestiae ut ut quas totam\nnostrum rerum est autem sunt rem eveniet architecto'}
Title field: sunt aut facere repellat provident occaecati excepturi optio reprehenderit


#### Explain:

response.json() converts JSON into Python dict/list.​

You can access fields like data["title"], data["body"], data["userId"].

Ask students to modify the URL: try /posts/2, /posts/3.

#### Part 6: Query parameters (simple URL example)
 Show that sometimes you pass extra filters in the URL.

Many APIs allow query parameters like ?userId=1 to filter data

In [3]:
import requests

url = "https://jsonplaceholder.typicode.com/posts"
params = {"userId": 1}

response = requests.get(url, params=params)
print("Final URL:", response.url)     # See ?userId=1
print("Status code:", response.status_code)

posts = response.json()
print("Number of posts for user 1:", len(posts))
print("First post:", posts[0])


Final URL: https://jsonplaceholder.typicode.com/posts?userId=1
Status code: 200
Number of posts for user 1: 10
First post: {'userId': 1, 'id': 1, 'title': 'sunt aut facere repellat provident occaecati excepturi optio reprehenderit', 'body': 'quia et suscipit\nsuscipit recusandae consequuntur expedita et cum\nreprehenderit molestiae ut ut quas totam\nnostrum rerum est autem sunt rem eveniet architecto'}


#### Part 7: Handling responses properly
Teach basic good habits:

Check status code before using data

In [4]:
import requests

url = "https://jsonplaceholder.typicode.com/posts/1"
response = requests.get(url)

if response.status_code == 200:
    data = response.json()
    print("Title:", data["title"])
else:
    print("Request failed with status:", response.status_code)


Title: sunt aut facere repellat provident occaecati excepturi optio reprehenderit


#### Simple error handling with try/except

In [5]:
import requests

url = "https://jsonplaceholder.typicode.com/posts/1"

try:
    response = requests.get(url, timeout=5)  # 5-second timeout
    response.raise_for_status()  # raises HTTPError for bad status
    data = response.json()
    print("Title:", data["title"])
except requests.exceptions.Timeout:
    print("Request timed out")
except requests.exceptions.HTTPError as e:
    print("HTTP error happened:", e)
except requests.exceptions.RequestException as e:
    print("Some other error:", e)


Title: sunt aut facere repellat provident occaecati excepturi optio reprehenderit


#### Simple POST example (demo only)

Even if you don’t go deep, show POST once so they see difference.

--> Concept:

POST sends data to the server, often in JSON format.​

In requests, you usually use requests.post(url, json=payload).​

Example with JSONPlaceholder (it accepts fake POSTs):

In [6]:
import requests

url = "https://jsonplaceholder.typicode.com/posts"
new_post = {
    "title": "My first API post",
    "body": "This is just a test.",
    "userId": 1
}

response = requests.post(url, json=new_post)

print("Status code:", response.status_code)
print("Response JSON:", response.json())


Status code: 201
Response JSON: {'title': 'My first API post', 'body': 'This is just a test.', 'userId': 1, 'id': 101}


| Code | Meaning      |
| ---- | ------------ |
| 200  | OK (success) |
| 201  | Created      |
| 400  | Bad Request  |
| 401  | Unauthorized |
| 404  | Not Found    |
| 500  | Server Error |
